### PR TITLE
Initial commit - main/users/books/transactions c files added

### PR DESCRIPTION
**Pull request by anniedroid**

### CODE REVIEW

File: books.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+enum BookTag { Biography, Fantasy, Fiction, Mystery, Romance, Sci_fi, Thriller, Young_adult };
+
+typedef struct Book
+{
+	long ISBN_num;
+	char* title;
+	int author_id;
+	enum BookStatus status;
+	enum BookTag tags[3];
```
Comment (saadali18) [inline books.c]: Dont limit to size of 3
Please confirm the syntax

File: books.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+enum BookStatus { inactive, active };
+
+enum BookTag { Biography, Fantasy, Fiction, Mystery, Romance, Sci_fi, Thriller, Young_adult };
```
Comment (saadali18) [inline books.c]: Maintain consistency while initializing enums

File: transactions.c [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include <stdio.h>
+
+typedef struct BookTransaction
+{
+	int book_uid;
+	char* check_out_date;
```
Comment (saadali18) [inline transactions.c]: How we are going to accommodate splitting datetime field 
e.g struct Date
    int month
    int day
    int year

File: users.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+enum Credential { Author, Customer, Librarian };
+
+enum UserStatus { inactive, active, delinquent };
```
Comment (saadali18) [inline users.c]: same as above
kindly lowercase your all enums

File: users.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+typedef struct User
+{
+	int user_id;
+	char* full_name;
+	char* login_name;
```
Comment (saadali18) [inline users.c]: Add a comment after login_name to call it a unique field

File: users.c [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	new_user->user_id = user_id;
+	new_user->full_name = _strdup(full_name);
+	new_user->login_name = _strdup(login_name);
+	new_user->password = _strdup(password); // Need to implement encryption
+	new_user->email = _strdup(email);
+	new_user->title = title;
+	new_user->status = status;
+	new_user->next = NULL; // Add new user to end of list
```
Comment (saadali18) [inline users.c]: create separate function for it `CreateUser`

File: users.c [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	char* email;
+	enum Credential title;
+	enum UserStatus status;
+	struct User* next;
+} User;
+
+
+User* user_list_head = NULL;
+
+void insertNewUser(int user_id, char* full_name, char* login_name, char* password, char* email, enum Credential title, enum UserStatus status, User** head)
```
Comment (saadali18) [inline users.c]: `insertUser`

File: users.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		current->next = new_user;
+	}
+	else
+	{
+		*head = new_user;
+	}
+}
+
+void printUserList(User* head)
```
Comment (saadali18) [inline users.c]: `printUsers`

please accommodate filters in your print function

File: users.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		case Librarian:
+			printf("Librarian\n");
+			break;
+		default:
+			break;
+		}
+		printf("Status: ");
+		switch (current->status)
+		{
+		case inactive:
```
Comment (saadali18) [inline users.c]: Do some research can we find out field exact names through enums instead of 0, 1, 2, 3

**Review by saadali18 — CHANGES_REQUESTED** [review]

File: books.c [hunk 10]
```diff
@@ -0,0 +1,162 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "users.h"
+
+enum book_status { book_inactive, book_active };
+
+enum book_tag { Biography, Fantasy, Fiction, Mystery, Romance, Sci_fi, Thriller, Young_adult };
```
Comment (saadali18) [inline books.c]: Please be consistent while defining enums

File: books.c [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+{
+	int book_uid; // Unique ID of each copy of same ISBN (original book)
+	long ISBN_num;
+	enum book_status status;
+	struct BookCopy* next;
+} BookCopy;
+
+Book* book_list_head = NULL;
+
+Book* createBook(long ISBN_num, char* title, int author_id, enum book_status status, enum book_tag* tags, int tag_count, int total_count, int in_stock_count, int likes)
```
Comment (saadali18) [inline books.c]: I think we don't need likes and in_stock_count

File: books.c [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+		current->next = book;
+	}
+	else
+	{
+		*head = book;
+	}
+}
+
+void searchBooksbyAuthor(int author_id)
+{
```
Comment (saadali18) [inline books.c]: I think we should change the parameter of this function

File: books.c [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+			printf("Title: %s\n", current->title);
+			switch (current->status)
+			{
+			case book_inactive:
+				printf("Book Status: Inactive\n");
+				break;
+			case book_active:
+				printf("Book Status: Active\n");
+				break;
+			default:
```
Comment (saadali18) [inline books.c]: I think we can create separate function for book status names

File: books.c [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+				case Sci_fi:
+					printf("Sci_fi\t");
+					break;
+				case Thriller:
+					printf("Thriller\t");
+					break;
+				case Young_adult:
+					printf("Young_adult\t");
+					break;
+				default:
```
Comment (saadali18) [inline books.c]: Same here

File: users.c [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+	char* email;
+	enum credential title;
+	enum user_status status;
+	struct User* next;
+} User;
+
+
+User* author_list_head = NULL;
+User* customer_list_head = NULL;
+User* librarian_list_head = NULL;
```
Comment (saadali18) [inline users.c]: I think we can combine them

File: users.c [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+		{
+			current = current->next;
+		}
+		current->next = user;
+	}
+	else
+	{
+		*head = user;
+	}
+}
```
Comment (saadali18) [inline users.c]: I think we can move this logic in utilities file

File: users.c [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+void printUsers(User* head)
```
Comment (saadali18) [inline users.c]: I think we can create searchUsers function, instead of printUsers

**Review by saadali18 — CHANGES_REQUESTED** [review]

File: users.c (hunk 16 above)
Comment (anniedroid) [reply users.c]: in progress

File: helpers.c [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+	return cred[title];
+}
+
+char* formatUserStatus(enum user_status status)
+{
+	char* user_s[] = { "Inactive", "Active", "Delinquent" };
+	return user_s[status];
+}
+
+void insertNode(void* node, void** head) // Attempt to use this as template
```
Comment (saadali18) [inline helpers.c]: Kindly try default parameters concept here

File: main.c [hunk 19]
```diff
@@ -0,0 +1,54 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "users.h"
+#include "helpers.h"
+
+int main()
+{
```
Comment (saadali18) [inline main.c]: Start building GUI as well

**Review by saadali18 — CHANGES_REQUESTED** [review]

File: books.c [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+	book->next = NULL;
+
+	return book;
+}
+
+void insertBook(Book* book, Book** head)
+{
+	if (*head)
+	{
+		Book* current = *head;
```
Comment (saadali18) [inline books.c]: if !(*head)
{ 
  *head= book;
   return;
}

File: books.c [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+{
+	if (!strcmp(toLowerCase(user_data), toLowerCase(search_data)))	return 1; // a match
+	return 0;
+}
+
+int filterbyLikes(int user_likes, int search_likes)
+{
+	if (user_likes >= search_likes)		return 1;
+	return 0;
+}
```
Comment (saadali18) [inline books.c]: 10 books (active ,  likes).


book* filterbyLikes(int user_likes, Book **head).  (10, head)
{
  Book *filteres_books = null
  current = head;
  while current:
          if current.likes ==  user_likes
               insert_book(current, filtered_books)
   return filtered_books
} 

book* filterbyStatus(int status, Book **head).  (10, head)
{
  Book *filteres_books = null
  current = head;
  while current:
          if current.status ==  status
               insert_book(current, filtered_books)
   return filtered_books
} 

parameter{
  likes: 10
   status: 10
}

Book* filterBooks(Book* parameter, Book** head)
{
    book *filtered_books = head;
     if paratemer->likes:  filtered_books = filterbyLikes(parameter->likes, filtered_books).  O(n)
     if parameter->status:  filtered_books = filterbyStatus(parameter->status, filtered_books) O(n)
     return filtered_books;
}

**Review by saadali18 — CHANGES_REQUESTED** [review]

File: books.c (hunk 21 above)
Comment (anniedroid) [reply books.c]: I found some issues with the filterBooks function code above we discussed in class:
I think: 
1) The Book* variables filter_book_on_likes, filter_book_on_status, filterbook_onID, etc - only defined within their respective if condition block and cannot be used to pass to the next if condition
2) Logical issues: the following "if" conditions are still dependent on the previous ones. For example: if parameters->status is not NULL (user passed in something to filter) but filter_book_on_like is NULL (we did'nt get anything back from last if statement), that "if filtered_books_on_likes && parameter->status" block will not run.

File: books.c [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+	while (current)
+	{
+		if (current->author_id == input_ID)
+			insertBook(current, filtered_result);
+		current = current->next;
+	}
+	return filtered_result;
+}
+
+Book* filterbyBookStatus(enum book_status input_status, Book** head) // will handle convert from user input string to enum in main
```
Comment (saadali18) [inline books.c]: Avoid unnecessary comments

File: books.h [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+	struct Book* next;
+} Book;
+
+typedef struct BookCopy
+{
+	int book_uid; // Unique ID of each copy of same ISBN (original book)
+	char* ISBN;
+	enum book_status status;
+	struct BookCopy* next;
+} BookCopy;
```
Comment (saadali18) [inline books.h]: Why do we need a BookCopy struct?

File: helpers.c [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+
+int isContain(char* data, char* keyword) // Check if a string contain the keyword input from user
+{
+	if (!data || !keyword)	return 0;
+
+	if (!strstr(toLowerCase(data), toLowerCase(keyword)))	return 0;
+	else	return 1;
+}
+
+int countDigits(long num)
```
Comment (saadali18) [inline helpers.c]: int length(long num)

File: helpers.c [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+	while (num > 0)
+	{
+		count++;
+		num = num / 10;
+	}
+
+	return count;
+}
+
+char* converttoString(long num)
```
Comment (saadali18) [inline helpers.c]: char *toString(long num)

File: books.c [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+	while (current)
+	{
+		if (!strcmp(current->ISBN, input_ISBN))
+			insertBook(current, filtered_result);
+		current = current->next;
+	}
+	return filtered_result;
+}
+
+Book* filterbyBookTitle(char* input_title, Book** head)
```
Comment (saadali18) [inline books.c]: kindly be consistent with function naming
Above you are using filterbyISBN, here you are using Book keyword as well

**Review by saadali18 — CHANGES_REQUESTED** [review]

**Comment by saadali18** [comment]
Key Points: 

Always search for buildin functions before creating custom functions
Be consistent with naming 
Avoid commented code

**Pull request closed without merging**